mailpoet/lib/API/MP/v1/Subscribers.php

```diff
@@ -84,61 +87,8 @@ public function subscribeToLists(
-    }
-
-    if (empty($subscriberId)) {
-      throw new APIException(__('A subscriber is required.', 'mailpoet'), APIException::SUBSCRIBER_NOT_EXISTS);
-    }
-
-    // throw exception when subscriber does not exist
-    $subscriber = null;
-    if (is_int($subscriberId) || (string)(int)$subscriberId === $subscriberId) {
-      $subscriber = $this->subscribersRepository->findOneById($subscriberId);
-    } else if (strlen(trim($subscriberId)) > 0) {
-      $subscriber = $this->subscribersRepository->findOneBy(['email' => $subscriberId]);
-    }
-
-    if (!$subscriber) {
-      throw new APIException(__('This subscriber does not exist.', 'mailpoet'), APIException::SUBSCRIBER_NOT_EXISTS);
-    }
-
-    // throw exception when none of the segments exist
-    $foundSegments = $this->segmentsRepository->findBy(['id' => $listIds]);
-    if (!$foundSegments) {
-      $exception = _n('This list does not exist.', 'These lists do not exist.', count($listIds), 'mailpoet');
-      throw new APIException($exception, APIException::LIST_NOT_EXISTS);
-    }
-
-    // throw exception when trying to subscribe to WP Users or WooCommerce Customers segments
-    $foundSegmentsIds = [];
-    foreach ($foundSegments as $foundSegment) {
-      if ($foundSegment->getType() === SegmentEntity::TYPE_WP_USERS) {
-        // translators: %d is the ID of the segment
-        throw new APIException(sprintf(__("Can't subscribe to a WordPress Users list with ID %d.", 'mailpoet'), $foundSegment->getId()), APIException::SUBSCRIBING_TO_WP_LIST_NOT_ALLOWED);
-      }
-      if ($foundSegment->getType() === SegmentEntity::TYPE_WC_USERS) {
-        // translators: %d is the ID of the segment
-        throw new APIException(sprintf(__("Can't subscribe to a WooCommerce Customers list with ID %d.", 'mailpoet'), $foundSegment->getId()), APIException::SUBSCRIBING_TO_WC_LIST_NOT_ALLOWED);
-      }
-      if ($foundSegment->getType() !== SegmentEntity::TYPE_DEFAULT) {
-        // translators: %d is the ID of the segment
-        throw new APIException(sprintf(__("Can't subscribe to a list with ID %d.", 'mailpoet'), $foundSegment->getId()), APIException::SUBSCRIBING_TO_LIST_NOT_ALLOWED);
-      }
-      $foundSegmentsIds[] = $foundSegment->getId();
-    }
-
-    // throw an exception when one or more segments do not exist
-    if (count($foundSegmentsIds) !== count($listIds)) {
-      $missingIds = array_values(array_diff($listIds, $foundSegmentsIds));
-      $exception = sprintf(
-        // translators: %s is a comma-separated list of list IDs.
-        _n('List with ID %s does not exist.', 'Lists with IDs %s do not exist.', count($missingIds), 'mailpoet'),
-        implode(', ', $missingIds)
-      );
-      throw new APIException(sprintf($exception, implode(', ', $missingIds)), APIException::LIST_NOT_EXISTS);
-    }
+    $subscriber = $this->findSubscriber($subscriberId);
+    $foundSegments = $this->getAndValidateSegments($listIds, self::CONTEXT_SUBSCRIBE);
```

Comment:
Do you think it might make sense to retrieve the segments before the subscriber just to be consistent with the old code? I doubt it really matters, but if someone made an API call with both an empty subscriber and no segment IDs, the previous code would return an error about requiring a segment while this would return an error about requiring a subscriber, right?

Reply:
I was thinking that it was not a big deal. We can't fix it simply by changing the order of those methods because, in that case, we would change the order of subsequent checks. So I guess I could extract the basic checks of inputs and call them before so that we have consistent order of checks. 
I can imagine some plugin might be using an error message for its logic, so to be safe, I'll rather fix it.